chain/chain/src/runtime/mod.rs

```diff
@@ -874,7 +885,16 @@ impl RuntimeAdapter for NightshadeRuntime {
                 storage_config.use_flat_storage,
             ),
         };
-        if checked_feature!("stable", StatelessValidationV0, PROTOCOL_VERSION)
+        let next_epoch_id =
+            self.epoch_manager.get_next_epoch_id_from_prev_block(&block.prev_block_hash)?;
+        let next_protocol_version =
+            self.epoch_manager.get_epoch_protocol_version(&next_epoch_id)?;
+
+        // We need to start recording reads if the stateless validation is
+        // enabled in the next epoch. We need to save the state transition data
+        // in the current epoch to be able to produce the state witness in the
+        // next epoch.
+        if checked_feature!("stable", StateWitnessSizeLimit, next_protocol_version)
```

Comment:
It'd be more proper to check both the current and next protocol versions. Maybe at some point in the future we will want to transition from stateless validation to something else and next epoch isn't going to have the feature, even though we still need to record things for the current epoch.

But for now it doesn't really matter, lgtm :+1: